src/games/strategy/common/swing/SwingComponents.java

```diff
@@ -91,17 +100,31 @@ public static void promptUser(final String title, final String message, final Ru
 
   public static void promptUser(final String title, final String message, final Runnable confirmedAction,
       final Runnable cancelAction) {
-    SwingUtilities.invokeLater(() -> {
-      int response = JOptionPane.showConfirmDialog(null, message, title,
-          JOptionPane.YES_NO_OPTION, JOptionPane.QUESTION_MESSAGE);
-      boolean result = response == JOptionPane.YES_OPTION;
-
-      if (result) {
-        confirmedAction.run();
-      } else {
-        cancelAction.run();
+
+    boolean showMessage = false;
+    synchronized (visiblePrompts) {
```

Comment:
If this action were launched from a button, double clicking it could have the threads race each other. Worst case a confirmation dialog pops up twice, which is not the end of the world either.


> Couldn't you have one thread run through the sync block then another flip the showMessage flag to false which would stomp on showing the first prompt?

No, the showMessage flag is method local, each thread has its own copy. If two threads were to pile up on the synchronized block, they would both have `false` for the value of `showMessage` Then when the next thread is allowed in the synchronized block, they will see that `visiblePromps` contains its message and will then fall through.